### DISCUSSION TITLE
How to enable query parameter validation for pydantic models

### DISCUSSION THREAD

**Discussion by maxispeicher** [fastapi/fastapi#10454] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

import uvicorn
from fastapi import Depends, FastAPI, Query
from pydantic import BaseModel

app = FastAPI()


class WorkingPageParams(BaseModel):
    page: int
    page_size: int

    def __init__(self, page: Annotated[int, Query(ge=1)] = 1, page_size: Annotated[int, Query(ge=1, le=500)] = 20):
        super().__init__(page=page, page_size=page_size)


class NotWorkingPageParams(BaseModel):
    page: Annotated[int, Query(ge=1)] = 1
    page_size: Annotated[int, Query(ge=1, le=500)] = 20


@app.get("/working")
def read_root(page_params: WorkingPageParams = Depends(WorkingPageParams)):
    return page_params


@app.get("/not-working")
def read_root(page_params: NotWorkingPageParams = Depends(NotWorkingPageParams)):
    return page_params


if __name__ == '__main__':
    uvicorn.run("main:app", host="127.0.0.1", reload=True)
```

### Description

Expected behavior would be that for both endpoints querying with e.g. `page_size=-1` returns a `422` status code with a descriptive error code.
However, when only using the pydantic model it seems like FastAPI is not properly "parsing" the model fields and doesn't identify the query parameters with their corresponding limits. The result of this is a `500` status code for the `not-working` endpoint.

Is there any way in which I can restructure the `BaseModel` without having to duplicate the field in the `__init__`?

### Operating System

Linux

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.10.12

**Comment by ebottos94** [comment, marked as answer]
In my opinion this is not correct way to implement Query Parameters : https://fastapi.tiangolo.com/tutorial/query-params-str-validations/

Anyway, this seems work : 

```python
from typing import Annotated

import uvicorn
from fastapi import Depends, FastAPI, Query
from fastapi.testclient import TestClient
from pydantic import BaseModel, Field

app = FastAPI()


class PageParams(BaseModel):
  page: int = 1
  page_size: int | None = Field(Query(default=20, ge=1, le=500))


@app.get("/query-params")
def read_root(page_params: PageParams = Depends()):
  return page_params

client = TestClient(app=app)

def test_query_params():
  response = client.get('/query-params?page_size=-1')
  assert response.status_code == 422
```

**Reply by maxispeicher** [reply]
Hey @ebottos94,

indeed this seems to work even though it leaves me a little bit puzzled why.

However, I'm not sure if it's not the intended behavior to implement query parameters. In the end the class use case for common query parameters is outlined here: https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/

But it's using a default class without any additional validation instead of a pydantic BaseModel. Not sure if this is intentionally or really due to some drawbacks which might not be limited to this case here.

**Comment by sehraramiz** [comment]
i had a similar issue, when using Depends for query params that uses a pydantic BaseModel object the swagger did not show the Field description **(Filter1 in example)**

this workaround with setting the defaut value of the model field as ```Query()``` works because of [this section](https://github.com/tiangolo/fastapi/blob/36c26677682c4245183912e00fc057c05cc6cf7a/fastapi/dependencies/utils.py#L378) of fastapi code (but i think there should be a better way)

i also wanted this behaviour to happen when using ```Annotated``` as type annotation of a pydantic model field but it does not work as expected.  **(Filter3 in example)**

is it ok to assume that at the end, the correct behaviour is that if we use Field as ```Annotated[str, Field()]``` or like ```field = Field()``` fastapi should use this FIeldInfo type values when analyzing api paramerers?

with this end behaviour in mind i changed some code to make the senario with ```Annotated[str, Query(description="foo")]``` to work.  **(Filter4 in example)**
the source of this issue is on [this section](https://github.com/tiangolo/fastapi/blob/0.105.0/fastapi/dependencies/utils.py#L208C5-L208C24) 
when we use inspect on a pydantic BaseModel the signature of the model is stripped out of Annotated metadata because of how pydantic handles it's signature(i assume). so i changed the code to replace the param evaluated annotation with it's original Annotated one.
ps: this does not solve this issue when using ```Field``` instead of ```Query```.  **(Filter3 in example)**

### example code:
```python
from typing import Annotated
from fastapi import FastAPI, Depends, Query
from pydantic import BaseModel, Field

app = FastAPI()


# not showing description in docs
class Filter1(BaseModel):
    size1: int = Field(description="filter 1")


# showing Query description (filter 2/1) in docs
class Filter2(BaseModel):
    size2: int = Field(
        default=Query(description="filter 2/1"), description="filter 2/2"
    )


# not showing description in docs
class Filter3(BaseModel):
    size3: Annotated[int, Field(description="filter 3")] = 100


# showing description in docs only when patch is applied
class Filter4(BaseModel):
    size4: Annotated[int, Query(description="filter 4")] = 100


# correctly showing description in body parameter is docs
class Filter5(BaseModel):
    size5: int = Field(description="filter 5")


@app.get("/")
def read_root(
    filter_body: Filter5,
    filter1: Filter1 = Depends(),
    filter2: Filter2 = Depends(),
    filter3: Filter3 = Depends(),
    filter4: Filter4 = Depends(),
):
    return {}
```

### experimental fix:
```diff
diff --git a/fastapi/dependencies/utils.py b/fastapi/dependencies/utils.py
index 4e88410a..37e6554c 100644
--- a/fastapi/dependencies/utils.py
+++ b/fastapi/dependencies/utils.py
@@ -208,15 +208,22 @@ def get_flat_params(dependant: Dependant) -> List[ModelField]:
 def get_typed_signature(call: Callable[..., Any]) -> inspect.Signature:
     signature = inspect.signature(call)
     globalns = getattr(call, "__globals__", {})
-    typed_params = [
-        inspect.Parameter(
-            name=param.name,
-            kind=param.kind,
-            default=param.default,
-            annotation=get_typed_annotation(param.annotation, globalns),
+    call_annotations = getattr(call, "__annotations__", {})
+    typed_params = []
+
+    for param in signature.parameters.values():
+        param_annotation = call_annotations.get(param.name)
+        if get_origin(param_annotation) is Annotated:
+            param = param.replace(annotation=param_annotation)
+        typed_params.append(
+            inspect.Parameter(
+                name=param.name,
+                kind=param.kind,
+                default=param.default,
+                annotation=get_typed_annotation(param.annotation, globalns),
+            )
         )
-        for param in signature.parameters.values()
-    ]
+
     typed_signature = inspect.Signature(typed_params)
     return typed_signature

```
![Screenshot 2023-12-21 at 21-48-55 FastAPI - Swagger UI](https://github.com/tiangolo/fastapi/assets/14166324/3aec38b9-47d0-499c-b06b-6c4c21ee4dce)
